File: src/example2.cc
#include "aeron-rust-wrapper/cxx_wrapper/example.h"
#include "aeron-rust-wrapper/cxx_wrapper/example2.h"
#include <iostream>


namespace example2{
Object2::Object2(const std::string & name):
m_name(name),
m_obj1(example::Object(name+"-obj1")),
m_obj2(example::Object(name+"-obj2"))
{ std::cout << "construct Object2-"<<m_name<< std::endl; }

//Object2::Object2(const std::string & name){
//m_name = name;
//m_obj1 = example::Object(name+"-obj1");
//m_obj2 = example::Object(name+"-obj2");
//std::cout << "construct Object2-"<<m_name<< std::endl;
//}


Object2::~Object2() { std::cout << "~Object2-"<<m_name << std::endl; }

void Object2::sayHi() const { std::cout << "Hi from Object2-"<<m_name << std::endl; }

//std::shared_ptr<Object2> createShared(const std::string & name) {
//  return std::make_shared<Object2>(Object2(name));
//}
//std::unique_ptr<Object2> createUnique(const std::string & name) {
//  return std::make_unique<Object2>(Object2(name));
//}

std::shared_ptr<Object2> createShared(const std::string & name) {
  return std::shared_ptr<Object2>(new Object2(name));
}
std::unique_ptr<Object2> createUnique(const std::string & name) {
  return std::unique_ptr<Object2>(new Object2(name));
}
}

File: src/example2.rs
use cxx::CxxString;
#[cxx::bridge(namespace = "example2")]
pub mod ffi {
    unsafe extern "C++" {

        #[namespace = "example"]
        type Object = crate::example::ffi::Object;

        include!("aeron-rust-wrapper/cxx_wrapper/example2.h");
        type Object2;

        fn createShared(name: &CxxString) -> SharedPtr<Object2>;
        fn createUnique(name: &CxxString) -> UniquePtr<Object2>;
        fn sayHi(self: &Object2);
        fn getObj1(self: &Object2) -> SharedPtr<Object>;
        fn getObj1Mut(self: Pin<&mut Object2>) -> SharedPtr<Object>;
        fn getObj2(self: &Object2) -> UniquePtr<Object>;
        fn getObj2Mut(self: Pin<&mut Object2>) -> UniquePtr<Object>;
    }
}